### ISSUE TITLE
1.6.9 breaks older version of Node

### ISSUE THREAD

**Issue by Mudrekh** [Atinux/schema-inspector#77]
@Atinux When upgrading dependencies from 1.6.8 to 1.6.9 (Async), this module breaks for older versions of Node.

While I understand the changes, to indicate a lack of support for Node versions, you need to update a major version to not mess with semver and default installation to take latest minor version. 

Ultimately 1.6.9 should have been published as 2.0.0 dropping support for < v8

**Comment by Atinux** [comment]
Hi @Mudrekh 

Thank you but I had to publish this release because of a security issue 😬 

Do you have the error message when running on node < v8?

Please mention me so I can get an email :)

**Comment by Mudrekh** [comment]
@Atinux Here is a snippet of the error. For this project, I just had to lock it at 1.6.8. Unfortunately, we are working on Node 4, but we are upgrading to 10 (and hopefully 12). It is just a slow process...

```/home/mudrekh/dev/scratch/node_modules/async/dist/async.js:52
    function apply(fn, ...args) {
                       ^^^
SyntaxError: Unexpected token ...
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:373:25)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at /home/mudrekh/dev/scratch/node_modules/schema-inspector/lib/schema-inspector.js:9:49
    at Object.<anonymous> (/home/mudrekh/dev/scratch/node_modules/schema-inspector/lib/schema-inspector.js:1577:3)
    at Module._compile (module.js:409:26)
```
Basically, that version of async uses spread operators which aren't available in Node 4.

**Comment by jcdesousa** [comment]
Also it stoped working for IE11, can be find here: https://caolan.github.io/async/v3/

"If you want to use Async in an older environment, (e.g. Node 4, IE11) you will have to transpile."

**Comment by Atinux** [comment]
Oh dammit, sadly I don't have time anymore to maintain this module. If you guys want to tackle this and become a maintainer, I am open to it 😄

**Comment by jcdesousa** [comment]
@Atinux From what i see there is no transpiling process, can I downgrade the async version to 2.6.2?

**Comment by Atinux** [comment]
Yes you can!

**Comment by mattwelke** [comment]
We actually noticed this requirement to transpile and just changed our build process to transpile using Webpack and Babel. We set the Babel config to include files in `node_modules`. I see nothing wrong with that requirement, to be honest. It makes sense that you'd ship modern code and let people transpile down to old code if their use case requires it.

**Comment by mattwelke** [comment]
This is on hold. I wasn't happy with my build since I was only able to get it to work with the `"useBuiltIns": "entry"` setting, which brings in the entire polyfill. The best way to do it would be with the `"useBuiltIns": "usage"` setting, but I wasn't able to get the build to work that way. It doesn't seem to bring in the polyfills correctly.

**Comment by jcdesousa** [comment]
@mattwelke I totally agree with the fact that we should change the build process to transpile using Webpack and Babel. 

But the patch upgrade from 1.6.8 to 1.6.9 introduced a breaking change that should no happen, that is why I was suggesting to downgrade the async to 2.6.2.

The change in the build process should happen on the major, or at least on the minor version.

**Comment by mattwelke** [comment]
It's funny you mention the breaking change idea. We had the same discussion at our work and we concluded that it wasn't necessary to bump the major version even if one of our dependencies included ES6 code, therefore breaking some build processes. We considered the semver version to relate to the API of schema-inspector, not its build process.

We were going to communicate to our consumers that if they weren't already testing their builds in browsers like IE, they should now because we're now aware of one of our dependencies shipping ES6+ code.

However, this is the open source world so we'll have to figure out our own way to move forward.

Is it appropriate for schema-inspector's maintainers to commit to testing builds of it in all browsers so that they can be made aware when one of its dependencies starts including ES6+? The library description describes it as meant to be used in the browser and on the server. So, to me, this seems reasonable.

Let's consider some consequences of going this route though. If we did, there would be a few ways to implement this:

- Increment the major version of schema-inspector and keep it on the latest version of async (and then, do we drop support for 2.x.x or maintain support for both major versions, doing multiple releases as necessary forever)
- Roll back async to when it used only ES5 code and keep it there, accepting the risk of not being able to update a dependency
- Replace usages of async code with our own ES5 code

**Comment by jcdesousa** [comment]
> We considered the semver version to relate to the API of schema-inspector, not its build process.

I see 🤔But you agree that a minor version should not break the out-of-box support of the  Node 4 version or older browsers, so it can be identified as breaking change. I think that we just need to align on the goal. 

Just consider the following, if we don't transpile and release ES5 code, Schema Inspector needs to reference that will drop the support for older browsers and node versions. I don't think that we want that.

My suggestion would be the following: 
* Release 1.6.10 - Rollback async to when it used only ES5 code with Async 2.6.3 (For now is fixed)
* Moving forward release 2.0.0, introduce a build process and transpile the Async Code that bundles the code (drop previous version support)
* Or stick with Async version 2 while we see that is still being supported.

**Comment by mattwelke** [comment]
> But you agree that a minor version should not break the out-of-box support of the Node 4 version or older browsers

Agreed. What makes sense to do right now is fix the semver violation. The steps you outlined make sense for that.

> My suggestion would be the following:
> 
> - Release 1.6.10 - Rollback async to when it used only ES5 code with Async 2.6.3 (For now is fixed)
> - Moving forward release 2.0.0, introduce a build process and transpile the Async Code that bundles the code (drop previous version support)
> - Or stick with Async version 2 while we see that is still being supported.

Agree with doing 1 and 2. 3 seems unwise. May as well fix it now.

Note that, like I said above, transpiling the code from the async dependency will not fix the problem for schema-inspector@2.x.x consumers who get it via NPM. It will only fix the issue for those who get the complete bundle build from the CDN. Let me know if you have any questions about how I'm coming to this conclusion. I can share example repos demonstrating how depending on schema-inspector 1.6.9 via NPM, even though it has no ES6+ code, still causes my own bundle to fail in EI (because of the transitive depenendencies with ES6+ code, async in this case).

**Comment by mattwelke** [comment]
@jcdesousa Looks like async@2.x.x will be supported long term with IE support: https://github.com/caolan/async/issues/1661

That might be the quickest solution here. Sounds like we can just change to using the 2.x.x branch of async, do a patch release, and not even worry about ES6+ code until further notice.

(side note, I got my Babel working to transpile it for IE without including the whole core-js polyfill 🎉 but I'm glad others won't have to go through that pain if we just use async@2.x.x )

**Comment by Atinux** [comment]
Fixed in #80 and released in https://github.com/Atinux/schema-inspector/releases/tag/v1.6.10

Thank you @mattwelke
